src/software/geom/angle_map.cpp

```diff
@@ -0,0 +1,79 @@
+#include "software/geom/angle_map.h"
+
+AngleMap::AngleMap(Angle top_angle, Angle bottom_angle, size_t max_num_obstacles)
+    : AngleMap(AngleSegment(top_angle, bottom_angle), max_num_obstacles)
+{
+}
+
+AngleMap::AngleMap(AngleSegment angle_seg, size_t max_num_obstacles)
+    : angle_seg(angle_seg)
+{
+    this->taken_angle_segments.reserve(max_num_obstacles);
+}
+
+const AngleSegment &AngleMap::getAngleSegment() const
+{
+    return this->angle_seg;
+}
+
+void AngleMap::addNonViableAngleSegment(AngleSegment &angle_seg)
+{
+    for (AngleSegment &taken_angle_seg : taken_angle_segments)
+    {
+        if (!(angle_seg.getAngleBottom() > taken_angle_seg.getAngleTop() ||
+              angle_seg.getAngleTop() < taken_angle_seg.getAngleBottom()))
+        {
+            taken_angle_seg.setAngleTop(
+                std::max(taken_angle_seg.getAngleTop(), angle_seg.getAngleTop()));
+            taken_angle_seg.setAngleBottom(
+                std::min(taken_angle_seg.getAngleBottom(), angle_seg.getAngleBottom()));
+            return;
+        }
+    }
+
+    this->taken_angle_segments.emplace_back(angle_seg);
+}
+
+AngleSegment AngleMap::getBiggestViableAngleSegment()
```

Comment:
nit can remove this-> throughout this function

Reply:
done